patches: update workaround for The Good Life (1452500)

popsUlfr opened the pull request:
Update to #44 

This rebases the workaround against the latest wine tree.

There's a problem with how the videos look though:

![1452500_20220221152724_1](https://user-images.githubusercontent.com/1030423/154978386-f5919587-2491-4db7-9bde-c5a19b129cef.png)

I'm pretty sure this is something out of my hands, the workaround simply replaces the `http:` url with a `file:` direct url.

Here's the ffprobe output of one of the video files if it could help:

```json
{
    "streams": [
        {
            "index": 0,
            "codec_name": "vp8",
            "codec_long_name": "On2 VP8",
            "profile": "0",
            "codec_type": "video",
            "codec_tag_string": "[0][0][0][0]",
            "codec_tag": "0x0000",
            "width": 1920,
            "height": 1080,
            "coded_width": 1920,
            "coded_height": 1080,
            "closed_captions": 0,
            "film_grain": 0,
            "has_b_frames": 0,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "16:9",
            "pix_fmt": "yuv420p",
            "level": -99,
            "color_range": "tv",
            "color_space": "bt709",
            "color_transfer": "bt709",
            "color_primaries": "bt709",
            "chroma_location": "left",
            "field_order": "progressive",
            "refs": 1,
            "r_frame_rate": "30/1",
            "avg_frame_rate": "30/1",
            "time_base": "1/1000",
            "start_pts": 0,
            "start_time": "0.000000",
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0
            },
            "tags": {
                "language": "eng",
                "HANDLER_NAME": "Mainconcept MP4 Video Media Handler",
                "VENDOR_ID": "[0][0][0][0]",
                "ENCODER": "Lavc59.1.100 libvpx",
                "DURATION": "00:00:06.000000000"
            }
        }
    ],
    "format": {
        "filename": "07c9c6b90acde468",
        "nb_streams": 1,
        "nb_programs": 0,
        "format_name": "matroska,webm",
        "format_long_name": "Matroska / WebM",
        "start_time": "0.000000",
        "duration": "6.000000",
        "size": "5986793",
        "bit_rate": "7982390",
        "probe_score": 100,
        "tags": {
            "COMPATIBLE_BRANDS": "isommp42",
            "MAJOR_BRAND": "mp42",
            "MINOR_VERSION": "0",
            "ENCODER": "Lavf59.2.100"
        }
    }
}
```

GloriousEggroll commented:
Thanks for the rebase!

the video issue can be fixed with `PROTON_DO_NOT_CREATE_DXGI_DEVICE_MANAGER=1 %command%`
A few other games are affected by this, see:

https://github.com/GloriousEggroll/proton-ge-custom/blob/b0731f252e495c123001cf9da6f011d0823c00e1/patches/proton/31-proton-mfplat-patches.patch#L2007

GloriousEggroll commented:
hmm, im guessing theres something else causing it to fail because the current patch applied on top of proton-ge is still failing/crashing.

popsUlfr commented:
> hmm, im guessing theres something else causing it to fail because the current patch applied on top of proton-ge is still failing/crashing.

What do you mean by it is failing/crashing ? For every game or just The Good Life ?
It's weird, it works for me though :(

Do you have maybe a log a log with `WINEDEBUG=trace+mfplat` ?

GloriousEggroll commented:
Luckily it no longer appears to be crashing other games like it was before (notably it was affecting Resident Evil 3), but the game itself is crashing:
![thegoodlife](https://user-images.githubusercontent.com/11287837/155087517-f9f53f41-445d-4b36-9b5e-6bd94298e54d.png)

Unfortunately it doesn't looke like it's even getting to mfplat before crashing. possibly crashing at kernelbase?

log attached:

[steam-1452500.log](https://github.com/GloriousEggroll/proton-ge-custom/files/8114855/steam-1452500.log)

popsUlfr commented:
That's not a lot to go with...

Did you try removing the compatdata folder ? (`~/.local/share/Steam/steamapps/compatdata/1452500`) I personally didn't need to to make it work.

Here is my local build the with the latest head at [a45fd85](https://github.com/GloriousEggroll/proton-ge-custom/commit/a45fd858556b0708374aadf6cac37e9d76721dc4) and my workaround patched in: 
- [proton-ge-thegoodlifeworkaround.tar.gz](https://mega.nz/file/JyRjWKwL#MV3AkCS7kk6mj1pRgPLAfh95Ke972Ej4ES4efcIucrM)

Here is just the `mfplat.dll` (64bit) from my build: [mfplat.dll.gz](https://github.com/GloriousEggroll/proton-ge-custom/files/8115231/mfplat.dll.gz)
 If I replace it with the `mfplat.dll` from your pre release build `Proton-7.2-GE-3-test-3` the game also works and shows the videos fine.

GloriousEggroll commented:
your build doesnt run for me, wineserver just hangs and the game never opens. replacing mfplat in my build also doesnt work, i get the same error as previously posted

popsUlfr commented:
That's all so weird. The only thing I could think of having an effect so early in the execution maybe is the fact that `wininet` is pulled in a linking dependency for the http requests ?

I don't really understand how the mfplat specific workaround that only fires in a certain condition is making is raising an unhandled page fault so early for you.

I'm on Arch Linux btw with apparmor enabled and a GTX 980Ti. Steam Beta and all the Steam Runtimes on the client_beta branch.

Here's my log file of a New Game through the intro videos:
[steam-1452500.log](https://github.com/GloriousEggroll/proton-ge-custom/files/8123252/steam-1452500.log)

```
======================
Proton: 1645520451 7.2-GE-2-test-2-8-ga45fd858
SteamGameId: 1452500
Command: ['/home/popsulfr/.local/share/Steam/steamapps/common/The Good Life/StandaloneWindows64.exe']
Options: {'forcelgadd'}
depot: 0.20220131.43
pressure-vessel: 0.20220131.0
scripts: v0.20211029.0-0-g2d89be4
soldier: 0.20220131.0 soldier 0.20220131.0
Kernel: Linux 5.16.10-zen1-1-zen #1 ZEN SMP PREEMPT Wed, 16 Feb 2022 19:35:21 +0000 x86_64
System WINEDEBUG: warn+mfplat
======================
fsync: up and running.
002c:fixme:winediag:LdrInitializeThunk wine-staging 7.2 is a testing version containing experimental patches.
002c:fixme:winediag:LdrInitializeThunk Please mention your exact version when filing bug reports on winehq.org.
Setting breakpad minidump AppID = 1452500
Steam_SetMinidumpSteamID:  Caching Steam ID:  76561197986026086 [API loaded no]
0148:err:steam:initialize_vr_data Could not load libopenvr_api.so.
0150:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_PERFORMANCE_INFORMATION
0150:fixme:imm:ImmReleaseContext (0000000000020080, 000000000037B810): stub
Fossilize INFO: Overriding serialization path: "/home/popsulfr/.local/share/Steam/steamapps/shadercache/1452500/fozpipelinesv6/steamapprun_pipeline_cache".
0150:fixme:wbemprox:client_security_SetBlanket 00000002419C1DE0, 0000000064EA6BD0, 10, 0, (null), 3, 3, 0000000000000000, 0
0150:fixme:wbemprox:client_security_Release 00000002419C1DE0
0150:fixme:pulse:pulse_channel_map_to_channel_mask Unhandled channel aux0
0150:fixme:pulse:pulse_channel_map_to_channel_mask Unhandled channel aux1
0150:fixme:pulse:pulse_channel_map_to_channel_mask Unhandled channel aux0
0150:fixme:pulse:pulse_channel_map_to_channel_mask Unhandled channel aux1
026c:fixme:kernelbase:AppPolicyGetThreadInitializationType FFFFFFFFFFFFFFFA, 000000007A6DFF50
026c:fixme:avrt:AvSetMmThreadCharacteristicsW (L"Audio",000000007A6DFEF8): stub
0150:fixme:win:RegisterTouchWindow (0000000000020080 00000000): stub
0124:fixme:imm:ImeSetActiveContext (0x4403d0, 0): stub
0124:fixme:imm:ImmReleaseContext (0000000000020054, 00000000004403D0): stub
01c0:fixme:mfplat:MFCreateDXGIDeviceManager stubbing out
02fc:fixme:combase:RoGetActivationFactory (L"Windows.System.DispatcherQueue", {00000035-0000-0000-c000-000000000046}, 00000001721BF960): semi-stub
02fc:err:combase:RoGetActivationFactory Failed to find library for L"Windows.System.DispatcherQueue"
02fc:fixme:combase:RoGetActivationFactory (L"Windows.System.DispatcherQueue", {00000035-0000-0000-c000-000000000046}, 00000001721BF510): semi-stub
02fc:err:combase:RoGetActivationFactory Failed to find library for L"Windows.System.DispatcherQueue"
winegstreamer warning: videobox0: not negotiated
winegstreamer warning: videobox0: ../src-gstreamer/libs/gst/base/gstbasetransform.c(1431): gst_base_transform_reconfigure_unlocked (): /GstBin:bin0/GstVideoBox:videobox0:
not negotiated
01c0:fixme:mfplat:src_reader_QueryInterface ({7b981cf0-560e-4116-9875-b099895f23d7}, 000000006C39F440)
0180:warn:mfplat:memory_buffer_QueryInterface Unsupported {7dc9d5f9-9ed9-44ec-9bbf-0600bb589fbb}.
...
```

popsUlfr commented:
Hi @GloriousEggroll 

I adjusted the workaround so it applies to the latest head of wine but I also needed to add the vp8 decoder to the ffmpeg configuration flag (also added vp9 just to be sure).
If you ever find the time could you see if it still crashes on your end ? I can still confirm that it works on my end.

The proper way to make the game's cutscenes work would be to implement the http: protocol scheme for transports in mfplat but that still looks very daunting to me.

Thanks and take care.

```patch
diff --git a/Makefile.in b/Makefile.in
index c5a776d1..81f475a9 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -596,7 +596,9 @@ FFMPEG_CONFIGURE_ARGS := \
        --enable-decoder=wmv2 \
        --enable-decoder=wmv1 \
        --enable-decoder=h264 \
-       --enable-decoder=aac
+       --enable-decoder=aac \
+       --enable-decoder=vp8 \
+       --enable-decoder=vp9
 
 $(eval $(call rules-source,ffmpeg,$(SRCDIR)/FFmpeg))
 $(eval $(call rules-configure,ffmpeg,32))
```
`thegoodlife-mfplat-http-scheme-workaround.patch`
```patch
From fab89e77c06e7059095dd47dc6728a49d6b417e3 Mon Sep 17 00:00:00 2001
From: Philipp Richter <richterphilipp.pops@gmail.com>
Date: Mon, 7 Mar 2022 15:43:45 +0100
Subject: [PATCH] mfplat: workaround for "The Good Life (1452500)" due to
 missing "http:" scheme implementation

"The Good Life (1452500)" wants to load its video files over http: but
that fails due to the missing implementation in wine.

This workaround pulls the Content-Length from the response headers and
attempts to match the right file in the game folder to serve it using
the "file:" scheme.

See: https://github.com/ValveSoftware/Proton/issues/5195
---
 dlls/mfplat/Makefile.in |   2 +-
 dlls/mfplat/main.c      | 205 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 206 insertions(+), 1 deletion(-)

diff --git a/dlls/mfplat/Makefile.in b/dlls/mfplat/Makefile.in
index 79af7650de6..d0196339042 100644
--- a/dlls/mfplat/Makefile.in
+++ b/dlls/mfplat/Makefile.in
@@ -1,7 +1,7 @@
 EXTRADEFS = -DWINE_NO_LONG_TYPES
 MODULE    = mfplat.dll
 IMPORTLIB = mfplat
-IMPORTS   = advapi32 ole32 mfuuid propsys rtworkq
+IMPORTS   = advapi32 ole32 mfuuid propsys rtworkq kernelbase wininet
 
 EXTRADLLFLAGS = -Wb,--prefer-native
 
diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index 0267381a44b..4894a8e9b3f 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -28,6 +28,7 @@
 #include "winbase.h"
 #include "winuser.h"
 #include "winreg.h"
+#include "wininet.h"
 
 #include "initguid.h"
 #include "rtworkq.h"
@@ -35,6 +36,7 @@
 #include "propsys.h"
 #include "d3d11.h"
 #include "uuids.h"
+#include "pathcch.h"
 
 #include "wine/debug.h"
 #include "wine/list.h"
@@ -6439,6 +6441,201 @@ static ULONG WINAPI source_resolver_Release(IMFSourceResolver *iface)
     return refcount;
 }
 
+/*
+ * Applies The Good Life (1452500) workaround for http: scheme urls
+ * Retrieves the Content-Length response header from the http request
+ * and matches it with the game files to find the direct path
+ * temp_buffer should be PATHCCH_MAX_CCH characters long for extended path support
+ * the new url will be placed into temp_buffer
+ */
+static BOOL thegoodlife_workaround_url_matcher(const WCHAR *url, WCHAR *temp_buffer, DWORD temp_buffer_size)
+{
+    static const WCHAR* acccept_types[] = { L"*/*", NULL };
+    URL_COMPONENTSW url_comp;
+    HINTERNET hinternet, hconnect, hrequest;
+    WCHAR content_length[24];
+    DWORD content_length_size;
+    WIN32_FIND_DATAW find_file_data;
+    HANDLE hfind;
+    DWORDLONG size, file_size;
+    content_length_size = sizeof(content_length);
+
+    memset(&url_comp, 0, sizeof(URL_COMPONENTSW));
+    url_comp.dwStructSize = sizeof(URL_COMPONENTSW);
+    url_comp.dwHostNameLength = 1;
+    url_comp.dwUserNameLength = 1;
+    url_comp.dwPasswordLength = 1;
+    url_comp.dwUrlPathLength = 1;
+    url_comp.dwExtraInfoLength = 1;
+
+    if (!(InternetCrackUrlW(url, lstrlenW(url), 0, &url_comp)))
+    {
+        WARN("InternetCrackUrlW failed: %u\n", GetLastError());
+        return FALSE;
+    }
+
+    /* write 0 terminated components into the temp_buffer */
+    if ((url_comp.dwHostNameLength + 1 +
+        url_comp.dwUserNameLength + 1 +
+        url_comp.dwPasswordLength + 1 +
+        url_comp.dwUrlPathLength + url_comp.dwExtraInfoLength + 1) > temp_buffer_size)
+    {
+        WARN("temp_buffer is too small\n");
+        return FALSE;
+    }
+
+    url_comp.lpszHostName = memcpy(temp_buffer, url_comp.lpszHostName, url_comp.dwHostNameLength * sizeof(WCHAR));
+    url_comp.lpszHostName[url_comp.dwHostNameLength] = 0;
+    url_comp.lpszUserName = memcpy(url_comp.lpszHostName + url_comp.dwHostNameLength + 1, url_comp.lpszUserName,
+        url_comp.dwUserNameLength * sizeof(WCHAR));
+    url_comp.lpszUserName[url_comp.dwUserNameLength] = 0;
+    url_comp.lpszPassword = memcpy(url_comp.lpszUserName + url_comp.dwUserNameLength + 1, url_comp.lpszPassword,
+        url_comp.dwPasswordLength * sizeof(WCHAR));
+    url_comp.lpszPassword[url_comp.dwPasswordLength] = 0;
+    url_comp.lpszUrlPath = memcpy(url_comp.lpszPassword + url_comp.dwPasswordLength + 1, url_comp.lpszUrlPath,
+        (url_comp.dwUrlPathLength + url_comp.dwExtraInfoLength) * sizeof(WCHAR));
+    url_comp.lpszUrlPath[url_comp.dwUrlPathLength + url_comp.dwExtraInfoLength] = 0;
+
+    if (!(hinternet = InternetOpenW(L"wininet", INTERNET_OPEN_TYPE_DIRECT, NULL, NULL, 0)))
+    {
+        WARN("InternetOpenW failed: %u\n", GetLastError());
+        return FALSE;
+    }
+
+    if (!(hconnect = InternetConnectW(hinternet, url_comp.lpszHostName, url_comp.nPort,
+        url_comp.lpszUserName, url_comp.lpszPassword, INTERNET_SERVICE_HTTP, 0, 0)))
+    {
+        WARN("InternetConnectW failed: %u\n", GetLastError());
+        InternetCloseHandle(hinternet);
+        return FALSE;
+    }
+
+    if (!(hrequest = HttpOpenRequestW(hconnect, L"GET", url_comp.lpszUrlPath,
+        NULL, NULL,
+        acccept_types,
+        INTERNET_FLAG_NO_CACHE_WRITE|INTERNET_FLAG_NO_COOKIES|INTERNET_FLAG_NO_UI|INTERNET_FLAG_RELOAD, 0)))
+    {
+        WARN("HttpOpenRequestW failed: %u\n", GetLastError());
+        InternetCloseHandle(hconnect);
+        InternetCloseHandle(hinternet);
+        return FALSE;
+    }
+
+    if (!HttpSendRequestW(hrequest, NULL, 0, NULL, 0))
+    {
+        WARN("HttpSendRequestW failed: %u\n", GetLastError());
+        InternetCloseHandle(hrequest);
+        InternetCloseHandle(hconnect);
+        InternetCloseHandle(hinternet);
+        return FALSE;
+    }
+
+    if (!HttpQueryInfoW(hrequest, HTTP_QUERY_CONTENT_LENGTH, content_length, &content_length_size, NULL))
+    {
+        WARN("HttpQueryInfoW failed: %u\n", GetLastError());
+        InternetCloseHandle(hrequest);
+        InternetCloseHandle(hconnect);
+        InternetCloseHandle(hinternet);
+        return FALSE;
+    }
+
+    InternetCloseHandle(hrequest);
+    InternetCloseHandle(hconnect);
+    InternetCloseHandle(hinternet);
+
+    if (!content_length_size)
+    {
+        WARN("Content-Length could not be retrieved\n");
+        return FALSE;
+    }
+
+    size = wcstoul(content_length, NULL, 10);
+
+    if (!size)
+        return FALSE;
+
+    /* Retrieve the current running application path */
+    if (!GetModuleFileNameW(NULL, temp_buffer, temp_buffer_size))
+    {
+        WARN("GetModuleFileNameW failed: %u\n", GetLastError());
+        return FALSE;
+    }
+
+    PathCchRemoveFileSpec(temp_buffer, temp_buffer_size);
+    PathCchAppendEx(temp_buffer, temp_buffer_size, L"StandaloneWindows64_Data\\StreamingAssets\\*",
+        PATHCCH_ALLOW_LONG_PATHS);
+
+    if ((hfind = FindFirstFileExW(temp_buffer, FindExInfoBasic, &find_file_data,
+        FindExSearchNameMatch, NULL, 0)) == INVALID_HANDLE_VALUE)
+    {
+        WARN("FindFirstFileExW failed: %u\n", GetLastError());
+        return FALSE;
+    }
+
+    do
+    {
+        if (find_file_data.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY)
+            continue;
+        if (find_file_data.nFileSizeHigh > 0)
+            file_size = (find_file_data.nFileSizeHigh * (MAXDWORD+1)) + find_file_data.nFileSizeLow;
+        else
+            file_size = find_file_data.nFileSizeLow;
+        if (file_size != size)
+            continue;
+        PathCchRemoveFileSpec(temp_buffer, temp_buffer_size);
+        PathCchAppendEx(temp_buffer, temp_buffer_size, find_file_data.cFileName,
+            PATHCCH_ALLOW_LONG_PATHS);
+        FindClose(hfind);
+        return TRUE;
+    }
+    while (FindNextFileW(hfind, &find_file_data));
+    if (GetLastError() != ERROR_NO_MORE_FILES)
+    {
+        WARN("FindNextFileW failed: %u\n", GetLastError());
+    }
+    FindClose(hfind);
+
+    return FALSE;
+}
+
+/*
+ * Tests if The Good Life (1452500) workaround needs to be applied
+ * temp_buffer should be PATHCCH_MAX_CCH characters long for extended path support
+ */
+static BOOL thegoodlife_workaround_scheme_check(const WCHAR *url, WCHAR *temp_buffer, DWORD temp_buffer_size)
+{
+    if (!url || lstrlenW(url) < 5 || memcmp(url, L"http:", sizeof(WCHAR) * 5))
+        return FALSE;
+
+    /* Retrieve the current running application path */
+    if (!GetModuleFileNameW(NULL, temp_buffer, temp_buffer_size))
+    {
+        WARN("GetModuleFileNameW failed: %u\n", GetLastError());
+        return FALSE;
+    }
+
+    PathCchRemoveFileSpec(temp_buffer, temp_buffer_size);
+    PathCchAppendEx(temp_buffer, temp_buffer_size, L"StandaloneWindows64_Data\\StreamingAssets",
+        PATHCCH_ALLOW_LONG_PATHS);
+    return PathFileExistsW(temp_buffer);
+}
+
+/* The Good Life (1452500) workaround macro */
+#define TGL_WORKAROUND(url) \
+    WCHAR __tgl_workaround_buffer[PATHCCH_MAX_CCH]; \
+    if (thegoodlife_workaround_scheme_check(url, __tgl_workaround_buffer, PATHCCH_MAX_CCH)) \
+    { \
+        TRACE("Applying The Good Life (1452500) workaround to %s.\n", debugstr_w(url)); \
+        if (!thegoodlife_workaround_url_matcher(url, __tgl_workaround_buffer, PATHCCH_MAX_CCH)) \
+        { \
+            WARN("Applying The Good Life (1452500) workaround to %s failed.\n", debugstr_w(url)); \
+        } \
+        else \
+        { \
+            url = __tgl_workaround_buffer; \
+        } \
+    }
+
 static HRESULT WINAPI source_resolver_CreateObjectFromURL(IMFSourceResolver *iface, const WCHAR *url,
         DWORD flags, IPropertyStore *props, MF_OBJECT_TYPE *obj_type, IUnknown **object)
 {
@@ -6447,6 +6644,8 @@ static HRESULT WINAPI source_resolver_CreateObjectFromURL(IMFSourceResolver *ifa
     IRtwqAsyncResult *result;
     RTWQASYNCRESULT *data;
     HRESULT hr;
+    /* The Good Life (1452500) workaround */
+    TGL_WORKAROUND(url)
 
     TRACE("%p, %s, %#x, %p, %p, %p.\n", iface, debugstr_w(url), flags, props, obj_type, object);
 
@@ -6488,6 +6687,8 @@ static HRESULT WINAPI source_resolver_CreateObjectFromByteStream(IMFSourceResolv
     IRtwqAsyncResult *result;
     RTWQASYNCRESULT *data;
     HRESULT hr;
+    /* The Good Life (1452500) workaround */
+    TGL_WORKAROUND(url)
 
     TRACE("%p, %p, %s, %#x, %p, %p, %p.\n", iface, stream, debugstr_w(url), flags, props, obj_type, object);
 
@@ -6529,6 +6730,8 @@ static HRESULT WINAPI source_resolver_BeginCreateObjectFromURL(IMFSourceResolver
     IUnknown *inner_cookie = NULL;
     IRtwqAsyncResult *result;
     HRESULT hr;
+    /* The Good Life (1452500) workaround */
+    TGL_WORKAROUND(url)
 
     TRACE("%p, %s, %#x, %p, %p, %p, %p.\n", iface, debugstr_w(url), flags, props, cancel_cookie, callback, state);
 
@@ -6573,6 +6776,8 @@ static HRESULT WINAPI source_resolver_BeginCreateObjectFromByteStream(IMFSourceR
     IUnknown *inner_cookie = NULL;
     IRtwqAsyncResult *result;
     HRESULT hr;
+    /* The Good Life (1452500) workaround */
+    TGL_WORKAROUND(url)
 
     TRACE("%p, %p, %s, %#x, %p, %p, %p, %p.\n", iface, stream, debugstr_w(url), flags, props, cancel_cookie,
             callback, state);
-- 
2.35.1

```
